version/version.go

```diff
@@ -17,14 +17,48 @@
 
 package version
 
+import (
+	"regexp"
+	"sync"
+)
+
+const (
+	defaultVersion = "0.0.0+unknown"
+)
+
 var (
 	// Package is filled at linking time
 	Package = "github.com/moby/buildkit"
 
 	// Version holds the complete version number. Filled in at linking time.
-	Version = "0.0.0+unknown"
+	Version = defaultVersion
 
 	// Revision is filled with the VCS (e.g. git) revision being used to build
 	// the program at linking time.
 	Revision = ""
 )
+
+var (
+	reSHA  *regexp.Regexp
+	reDev  *regexp.Regexp
+	reOnce sync.Once
+)
+
+func UserAgent() string {
+	version := Version
+
+	reOnce.Do(func() {
+		reSHA = regexp.MustCompile(`^[0-9a-f]{6,8}$`)
+		reDev = regexp.MustCompile(`^(v[0-9]+\.[0-9]+\.[0-9]+)-[0-9]+-g[0-9a-f]{6,10}(?:\.m)?$`)
+	})
+
+	if reSHA.Match([]byte(version)) {
+		version = defaultVersion
+	}
+
+	if matchs := reDev.FindAllStringSubmatch(version, 1); len(matchs) > 0 {
+		version = matchs[0][1] + "-dev"
+	}
+
+	return "buildkit/" + version
```

Comment:
Could we more info, that could be useful:

```go
return fmt.Sprintf("buildkit/%s go/%s %s/%s", version, runtime.Version()[2:], runtime.GOOS, runtime.GOARCH)
```

Reply:
I'm not sure. Go version can be implied from the main version. OS/Arch can be useful but we need to draw the line somewhere. The previous containerd string used the same formatting with only version.